chips/nrf52840/src/interrupt_service.rs

```diff
@@ -1,25 +1,39 @@
-use crate::gpio;
-use crate::peripheral_interrupts;
-use nrf52::interrupt_service::InterruptService;
+use crate::deferred_call_tasks::DeferredCallTask;
+use nrf52::chip::Nrf52DefaultPeripherals;
 
-pub struct Nrf52840InterruptService {
-    nrf52: nrf52::interrupt_service::Nrf52InterruptService<'static>,
+/// This struct, when initialized, instantiates all peripheral drivers for the nrf52840.
+/// If a board wishes to use only a subset of these peripherals, this
+/// should not be used or imported, and a modified version should be
+/// constructed manually in main.rs.
+//create all base nrf52 peripherals
+pub struct Nrf52840DefaultPeripherals<'a> {
+    pub nrf52_base: Nrf52DefaultPeripherals<'a>,
+    pub usbd: crate::usbd::Usbd<'a>,
 }
 
-impl Nrf52840InterruptService {
-    pub unsafe fn new() -> Nrf52840InterruptService {
-        Nrf52840InterruptService {
-            nrf52: nrf52::interrupt_service::Nrf52InterruptService::new(&gpio::PORT),
+impl<'a> Nrf52840DefaultPeripherals<'a> {
+    pub fn new(ppi: &'a crate::ppi::Ppi) -> Self {
+        Self {
+            nrf52_base: unsafe { Nrf52DefaultPeripherals::new(&crate::gpio::PORT, ppi) },
+            usbd: crate::usbd::Usbd::new(),
         }
     }
+    // Necessary for setting up circular dependencies
+    pub fn init(&'a self) {
+        self.nrf52_base.pwr_clk.set_usb_client(&self.usbd);
+        self.usbd.set_power_ref(&self.nrf52_base.pwr_clk);
```

Comment:
This is fine, but I wonder whether these two calls cannot be bundled in one?

Reply:
It could be done with a single (non-method) function, but would require changing the visibility of fields of one of the structs. I think this is more in line with how we set up circular dependencies elsewhere anyways.